### PR TITLE
cluster: reject the split request of a joint state region

### PR DESCRIPTION
**Pull request by HunDunDM**
Signed-off-by: Zheng Xiangsheng <hundundm@gmail.com>

### What is changed and how it works?

* If a region is in joint state, its split region request should be rejected.

### Check List

Tests

- Unit test

### Release note

- No release note

### CODE REVIEW

**Review by NingLin-P — APPROVED** [review]

File: server/cluster/cluster_worker.go [hunk 1]
```diff
@@ -45,7 +45,7 @@ func (c *RaftCluster) HandleRegionHeartbeat(region *core.RegionInfo) error {
 // HandleAskSplit handles the split request.
 func (c *RaftCluster) HandleAskSplit(request *pdpb.AskSplitRequest) (*pdpb.AskSplitResponse, error) {
 	reqRegion := request.GetRegion()
-	err := c.ValidRequestRegion(reqRegion)
+	err := c.ValidRequestSplitRegion(reqRegion)
```
Comment (rleungx) [inline server/cluster/cluster_worker.go:48]: Does `HandleAskBatchSplit` need change?

File: server/cluster/cluster_worker.go [hunk 2]
```diff
@@ -77,13 +77,17 @@ func (c *RaftCluster) HandleAskSplit(request *pdpb.AskSplitRequest) (*pdpb.AskSp
 	return split, nil
 }
 
-// ValidRequestRegion is used to decide if the region is valid.
-func (c *RaftCluster) ValidRequestRegion(reqRegion *metapb.Region) error {
+// ValidRequestSplitRegion is used to decide if the region is valid.
+func (c *RaftCluster) ValidRequestSplitRegion(reqRegion *metapb.Region) error {
```
Comment (rleungx) [inline server/cluster/cluster_worker.go:81]: Does the region size become larger if it is a hot write region?

File: server/cluster/cluster_worker.go [hunk 3]
```diff
@@ -77,13 +77,17 @@ func (c *RaftCluster) HandleAskSplit(request *pdpb.AskSplitRequest) (*pdpb.AskSp
...
 	startKey := reqRegion.GetStartKey()
 	region := c.GetRegionByKey(startKey)
 	if region == nil {
 		return errors.Errorf("region not found, request region: %v", logutil.RedactStringer(core.RegionToHexMeta(reqRegion)))
 	}
+	// If the region is in joint state, then returns an error.
+	if core.IsInJointState(reqRegion.Peers...) {
+		return errors.Errorf("region is in joint state, request region: %v", logutil.RedactStringer(core.RegionToHexMeta(reqRegion)))
```
Comment (JmPotato) [inline server/cluster/cluster_worker.go:89]: Why are these errors not defined through the `github.com/pingcap/errors` package?

Comment (HunDunDM) [reply server/cluster/cluster_worker.go:89]: These err need to be returned through GRPC.

File: server/cluster/cluster_worker.go (hunk 2 above)
Comment (HunDunDM) [reply server/cluster/cluster_worker.go:81]: If such a bug occurs, it needs to be resolved. Currently, this limitation will be judged in both PD and TiKV.

File: tests/server/cluster/cluster_work_test.go [hunk 4]
```diff
@@ -57,38 +56,91 @@ func (s *clusterWorkerTestSuite) TestValidRequestRegion(c *C) {
...
+		&metapb.Region{Id: 2, StartKey: []byte("a"), EndKey: []byte("b"), Peers: peers},
+		NotNil,
+	}, {
+		&metapb.Region{Id: 1, StartKey: []byte(""), EndKey: []byte("a"), Peers: peers, RegionEpoch: &metapb.RegionEpoch{ConfVer: 1, Version: 2}},
+		NotNil,
+	}, {
+		&metapb.Region{Id: 1, StartKey: []byte(""), EndKey: []byte("a"), Peers: peers, RegionEpoch: &metapb.RegionEpoch{ConfVer: 2, Version: 1}},
+		NotNil,
+	}, {
+		&metapb.Region{Id: 1, StartKey: []byte(""), EndKey: []byte("a"), Peers: peers, RegionEpoch: &metapb.RegionEpoch{ConfVer: 1, Version: 3}},
```
Comment (rleungx) [inline tests/server/cluster/cluster_work_test.go:97]: What is the difference between this line and line 91?

Comment (HunDunDM) [reply tests/server/cluster/cluster_work_test.go:97]: `Version: 3`

**Review by lhy1024 — APPROVED** [review]

**Review by nolouch — APPROVED** [review]
lgtm

**Comment by nolouch** [comment]
Does this PR need?

**Comment by JmPotato** [comment]
@HunDunDM Do we still need to merge this?

**Comment by disksing** [comment]
Can it be checked on TiKV side? I think PD's state can be not the most updated. For example when PD has recently changed leader.

**Pull request closed without merging**